### PR TITLE
RHOAIENG-68237: Remove gate-filtered tests from JUnit output to eliminate false "skipped" reports

### PR DESCRIPTION
**Pull request by heorhiig**
## Description
I added a small filter so tag-gate skips (like `Skipping test: passed tag:`) are removed from the e2e JUnit XML before CI picks it up.

Jira: https://redhat.atlassian.net/browse/RHOAIENG-68237

What I did:
- added `FilterGateSkippedTests` in `test-retry` formatter
- added CLI command `test-retry filter-gate-skips`
- hooked it into `run_e2e_tests.sh` after gotestsum / test-retry finishes

Real skips and failures stay in the report.

## How Has This Been Tested?
- `go test ./pkg/cli/ ./pkg/formatter/` in `cmd/test-retry`
- `bash -n tests/e2e/scripts/run_e2e_tests.sh`
- manually ran `test-retry filter-gate-skips --junit` on a sample XML

## Merge criteria
- [x] You have read the [contributors guide](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/CONTRIBUTING.md).
- [x] Commit messages are meaningful - have a clear and concise summary and detailed explanation of what was changed and why.
- [x] Pull Request contains a description of the solution, a link to the JIRA issue, and to any dependent or related Pull Request.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work
- [ ] The developer has run the integration test pipeline and verified that it passed successfully
- [ ] New RELATED_IMAGE mappings are listed in ODH-Build-Config and RHOAI-Build-Config (CI validates names against build-config repos). Include links to build-config PRs in the description.

### E2E test suite update requirement
- [x] Skip requirement to update E2E test suite for this PR

#### E2E update requirement opt-out justification
This only post-processes the JUnit report in CI tooling. It does not change operator behavior or the e2e suite itself.

## Summary of Updates

- **New Features**
  - Added a `filter-gate-skips` command to remove gate-skipped test cases from JUnit XML reports.
  - Automatically processes available reports after end-to-end test runs.

- **Bug Fixes**
  - Preserves other test results while updating report counts.
  - Supports standard and standalone JUnit suite formats.
  - Preserves the original test-run exit status when filtering succeeds or fails.

- **Tests**
  - Added coverage for filtering behavior, edge cases, XML formats, and error handling.

### CODE REVIEW

**Review by macgregor — CHANGES_REQUESTED** [review]
## Question: Is this solving a problem that exists in CI?

The Prow CI jobs (`opendatahub-operator-e2e`, `opendatahub-operator-rhoai-e2e`) both run `make e2e-test`, which uses `test-retry e2e` directly — not `run_e2e_tests.sh`. test-retry's parser ([`parser.go` lines 155-175](https://github.com/opendatahub-io/opendatahub-operator/blob/main/cmd/test-retry/pkg/parser/parser.go#L155-L175)) only collects `pkg.Failed` and `pkg.Passed`, never `pkg.Skipped`, and `TestResult` has [no `SkippedTest` field](https://github.com/opendatahub-io/opendatahub-operator/blob/main/cmd/test-retry/pkg/types/types.go#L31-L33) — so gate-skip noise shouldn't appear in Prow artifacts. This filter only touches `run_e2e_tests.sh` (the Docker entrypoint), which defaults to `USE_TEST_RETRY=false` (gotestsum). Which CI system / job is producing the noisy JUnit artifacts that motivated this?

---

## Implementation feedback (if the code is still wanted):

### 1. XML shape mismatch breaks the test-retry branch (blocking)

The filter expects `<testsuites>` root. test-retry writes `<testsuite>` ([`junit.go` line 18](https://github.com/opendatahub-io/opendatahub-operator/blob/main/cmd/test-retry/pkg/formatter/junit.go#L18): `XMLName xml.Name \x60xml:"testsuite"\x60`). Running the filter on both branches in `run_e2e_tests.sh` will error on the test-retry path, and `set -euo pipefail` means that error replaces the real test exit code. Either only run the filter in the gotestsum branch, or make the filter handle both XML shapes.

### 2. Shell error handling — fail open (blocking)

Even on the gotestsum path, if the filter fails for any reason, `set -euo pipefail` kills the script and the real test exit code is lost. The filter call should fail open:

```bash
test-retry filter-gate-skips --junit "$JUNIT_FILE" || {
    echo "Warning: gate-skip filter failed; leaving ${JUNIT_FILE} unfiltered" >&2
    return 0
}
```

A cosmetic post-processing step should never mask real test results.

### 3. Shared const for skip prefix (blocking)

`"Skipping test: passed tag:"` is hardcoded in both [`test_tag_test.go:29`](https://github.com/opendatahub-io/opendatahub-operator/blob/main/tests/e2e/test_tag_test.go#L29) (the `Skipf` call) and `filter_gate_skips.go:13`. If someone changes the format string, the filter silently stops matching. These should share a const — suggest a small shared package (e.g. `pkg/test/tagging/`) that both the test code and the formatter import.

### 4. Duplicate JUnit XML structs (request changes)

`formatter/` now has two parallel type hierarchies for the same XML format — `TestSuite`/`TestCase`/`Failure` in [`junit.go`](https://github.com/opendatahub-io/opendatahub-operator/blob/main/cmd/test-retry/pkg/formatter/junit.go#L17-L43) and `filterTestSuites`/`filterTestSuite`/etc. in the new file. The existing types lack `Skipped` and the `<testsuites>` envelope, but extending them is better than maintaining two sets that will drift.

### 5. Atomic file write (suggestion)

`FilterGateSkippedTestsFile` overwrites in place. Write to a temp file + `os.Rename` for safety — cheap insurance against losing the only copy of the JUnit report on a partial write.

File: cmd/test-retry/pkg/formatter/filter_gate_skips.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+import (
+	"encoding/xml"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+const gateSkipPrefix = "Skipping test: passed tag:"
```
Comment (den-rgb) [inline cmd/test-retry/pkg/formatter/filter_gate_skips.go]: import tagging.GateSkipPrefix from tests/tagging instead of duplicating the literal.

File: tests/e2e/scripts/run_e2e_tests.sh
Comment (den-rgb) [inline tests/e2e/scripts/run_e2e_tests.sh:161]: running the filter on the test-retry branch is a no-op (test-retry never emits <skipped>) but changes the XML root from <testsuite> to <testsuites>

**Review by KevFan — APPROVED** [review]

**Review by KevFan — APPROVED** [review]

**Comment by KevFan** [comment]
I think the merging of this was prevented multiple times even when it was approved due to this PR having the `do-not-merge/hold` label on it. Looks like the `Openshift-ci` bot added this. Do we know why this was added? 🤔 \cc @macgregor

**Comment by asmigala** [comment]
@KevFan it was added because the test kept failing: https://github.com/opendatahub-io/opendatahub-operator/pull/3818#issuecomment-5543962958

**Comment by KevFan** [comment]
Thanks! @asmigala 👍 

/lgtm
